torch/storage.py

```diff
@@ -156,10 +172,483 @@ def _new_shared(cls, size):
+                assert len(kwargs) == 2, (
+                    "Only 'wrap_storage' and 'dtype' should be given, but got: "
+                    f"{kwargs}")
+                dtype = kwargs['dtype']
+                assert isinstance(dtype, torch.dtype)
+                self.dtype = dtype
+
+            else:
+                assert hasattr(self, 'dtype')
+                assert len(kwargs) == 1, (
+                    f"Only 'wrap_storage' should be given, but got: {kwargs.keys()}")
+                dtype = self.dtype
+
+            storage = kwargs['wrap_storage']
+
+            if not isinstance(storage, (torch.ByteStorage, torch.cuda.ByteStorage)):
+                raise TypeError(arg_error_msg)
+            if type(self) != TypedStorage and storage.__module__ != self.__module__:
+                raise TypeError((
+                    arg_error_msg +
+                    f'\n`storage` `module {storage.__module__}` does not match '
+                    f'module of {type(self)}'))
+            self._storage = storage
+
+        else:
+            assert type(self) != TypedStorage, (
+                "Calling __init__ this way is only supported in TypedStorage's "
+                "child classes. TypedStorage can only be directly instantiated "
+                "when kwargs 'wrap_storage' and 'dtype' are given.")
+
+            assert len(kwargs) == 0, "invalid keyword arguments"
+
+            def isint(x):
+                try:
+                    int(x)
+                except TypeError:
+                    return False
+                return True
+
+            if len(args) == 0:
+                self._storage = eval(self.__module__).ByteStorage()
+
+            elif len(args) == 1 and isint(args[0]):
+                self._storage = eval(self.__module__).ByteStorage(int(args[0]) * self.element_size())
+
+            elif len(args) == 1 and isinstance(args[0], collections.abc.Sequence):
+                if self.dtype in [torch.quint8, torch.quint4x2, torch.qint32, torch.qint8]:
+                    interpret_dtypes = {
+                        torch.quint8: torch.uint8,
+                        torch.quint4x2: torch.uint8,
+                        torch.qint32: torch.int32,
+                        torch.qint8: torch.int8
+                    }
+                    tmp_tensor = torch.tensor(
+                        args[0],
+                        dtype=interpret_dtypes[self.dtype],
+                        device='cuda' if self.is_cuda else 'cpu')
+
+                else:
+                    tmp_tensor = torch.tensor(
+                        args[0],
+                        dtype=self.dtype,
+                        device='cuda' if self.is_cuda else 'cpu')
+
+                self._storage = tmp_tensor.storage()._untyped()
+
+            else:
+                raise TypeError(arg_error_msg)
+
+    @property
+    def is_cuda(self):
+        return eval(self.__module__) is torch.cuda
```

Comment:
I don't like this. `self._storage` should know its device and we should be able to query this from there